draft-ietf-tls-extended-key-update.md

```diff
@@ -764,26 +764,36 @@ SSLKEYLOGFILE secrets including past iterations of `CLIENT_TRAFFIC_SECRET_`,
 
 Protocols such as DTLS-SRTP and DTLS-over-SCTP rely on TLS or DTLS for
 key establishment, but reuse portions of the derived keying material for
-their own specific purposes.These protocols use the TLS exporter defined
+their own specific purposes. These protocols use the TLS exporter defined
 in {{Section 7.5 of TLS}}.
 
 Once the Extended Key Update mechanism is complete, such protocols would
-need to use the newly derived key to generate Exported Keying Material
+need to use the newly derived exporter secret to generate Exported Keying Material
 (EKM) to protect packets. The "sk" derived in the {{key_update}} will be
 used as the "Secret" in the exporter function, defined in
 {{Section 7.5 of TLS}}, to generate EKM, ensuring that
 the exported keying material is aligned with the updated security context.
 
-When the exporter master secret is updated following a successful Extended Key Update,
-the TLS/DTLS implementation will have to notify the application that a
-new exporter secret is available.
+When a new exporter secret becomes active following a successful Extended
+Key Update, the TLS or DTLS implementation would have to provide an
+asynchronous notification to the application indicating that:
+
+* A new epoch has become active; and
+
+* The corresponding EKM has been derived using the exporter construction
+  discussed above, together with the label and context value.
+
+Delivering the derived EKM in this notification allows applications that
+depend on exporter-based keying material to install new application-layer
+keys in synchronization with the epoch transition.
 
 To prevent desynchronization, the application will have to retain both the
 previous and the newly derived exporter secrets for a short period. For TLS,
 the previous exporter secret would be discarded once data derived from the
-new exporter has been successfully processed. For DTLS, the previous exporter
-secret needs to be retained until the retention timer expires, to allow for
-processing of packets that may arrive out of order.  The retention policy
+new exporter has been successfully processed, and no records protected with
+the old exporter secret are expected to arrive. For DTLS, the previous exporter
+secret needs to be retained until the retention timer expires for the prior epoch,
+to allow for processing of packets that may arrive out of order.  The retention policy
```

Comment:
For DTLS, processing out-of-order packets is not sufficient for applications that have some reliable application-level message that uses exporters. Even without re-ordering, just packet loss, such message can arrive well after epoch transition, possibly without any signs that such message is coming until it arrives.

Worse, since the message must be associated with correct epoch at DTLS level, it is never safe to free the epoch. And since there are only two epoch signaling bits in DTLS 1.3, once the fourth EKU happens, things go south.

This problem does not exist in base DTLS 1.3, since there is never need to associate message with epoch, so re-sending a message can freely use the newest epoch.

Reply:
DTLS 1.3 does not limit epochs to two bits, see https://datatracker.ietf.org/doc/rfc9147/. 

In DTLS-SRTP, SRTP packets are sent directly over RTP, not within DTLS. The exporter is used only to derive SRTP master keys, and replay-window handling in SRTP ensures correct processing of out-of-order packets.

DTLS-over-SCTP provides reliable, in-order delivery of DTLS records, so Extended Key Update and its ACKs are always processed before any data encrypted with the new epoch. This eliminates the synchronization issue in that context as well.

Reply:
I don’t agree with your interpretation, but to save everyone’s time, could you please propose specific text that you think would address your concerns?
That way, we can review it directly and see whether it aligns with the intent of the draft.